server/src/language-service/documentSymbolProvider.ts

```diff
@@ -0,0 +1,155 @@
+import {
+  HierarchySettings,
+  defaultHierarchySettings,
+  getHierarchyLevel,
+} from "../util";
+
+interface SepticSymbol {
+  symbol: DocumentSymbol;
+  level: number;
+  parent: SepticSymbol | undefined;
+}
+
+export class DocumentSymbolProvider {
+  private readonly cnfgProvider: ISepticConfigProvider;
+  private readonly settingsManager: SettingsManager;
+
+  constructor(
+    cnfgProvider: ISepticConfigProvider,
+    settingsManager: SettingsManager
+  ) {
+    this.cnfgProvider = cnfgProvider;
+    this.settingsManager = settingsManager;
+  }
+
+  public provideDocumentSymbols(
+    document: ITextDocument,
+    token: CancellationToken | undefined
+  ): DocumentSymbol[] {
+    const cnfg = this.cnfgProvider.get(document.uri);
+    if (!cnfg) {
+      return [];
+    }
+    const settings =
+      this.settingsManager.getSettings()?.hierarchy ?? defaultHierarchySettings;
+    return getDocumentSymbols(document, cnfg, settings);
+  }
+}
+
+export function getDocumentSymbols(
+  doc: ITextDocument,
+  cnfg: SepticCnfg,
+  settings: HierarchySettings
+) {
+  let symbols = cnfg.objects.map((obj) => {
+    return createSepticSymbol(obj, doc, settings);
+  });
+
+  let root = {
+    level: -1,
+    parent: undefined,
+    symbol: dummySymbol(),
+  };
+
+  buildTree(root, symbols);
+
+  root.symbol.children?.forEach((child) => {
+    updateRange(child);
+  });
+
+  return root.symbol.children!;
+}
+
+function buildTree(parent: SepticSymbol, symbols: SepticSymbol[]) {
+  if (!symbols.length) {
+    return;
+  }
+
+  let symbol = symbols[0];
+
+  while (parent && symbol.level <= parent.level) {
+    parent = parent.parent!;
+  }
+
+  parent.symbol.children?.push(symbol.symbol);
+  symbol.parent = parent;
+
+  buildTree(symbol, symbols.slice(1));
+}
+
+function updateRange(parent: DocumentSymbol) {
+  if (!parent.children?.length) {
+    return;
+  }
+  parent.children?.forEach((child) => {
+    updateRange(child);
+  });
+
+  let end = parent.children[parent.children.length - 1].range.end;
+  parent.range.end = end;
+  parent.selectionRange.end = end;
+}
+
+function createSepticSymbol(
+  obj: SepticObject,
+  doc: ITextDocument,
+  settings: HierarchySettings
+): SepticSymbol {
+  let name = obj.name + ": " + obj.variable?.name;
+  let symbolKind = getSymbolKind(obj.name);
+  let range = {
+    start: doc.positionAt(obj.start),
+    end: doc.positionAt(obj.end),
+  };
+  let level = getHierarchyLevel(obj, settings);
+  return {
+    symbol: DocumentSymbol.create(
+      name,
+      undefined,
+      symbolKind,
+      range,
+      range,
+      []
+    ),
+    level: level,
+    parent: undefined,
+  };
+}
+
+function dummySymbol(): DocumentSymbol {
+  let range = {
+    start: { character: 0, line: 0 },
+    end: { character: 0, line: 0 },
+  };
+  return DocumentSymbol.create(
+    "",
+    undefined,
+    SymbolKind.Array,
+    range,
+    range,
+    []
+  );
+}
+
+function getSymbolKind(type: string) {
```

Comment:
Should we consider putting all object types in a structured config file (yaml?) where all sorts of information can be extracted? E.g. folding level, symbolkind, possible parent/child relationships. This can of course be done at a later time, but I think it'll be worth it instead of hard-coding object names throughout the code files.

Reply:
Agree. Can start with a basic structure that provide hierarchy level, symbolkind etc. and expand as needed.

Reply:
Will be updated in #105